### PR TITLE
remove default priority cache in Priority admission controller

### PR DESCRIPTION
**Pull request by wackxu**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes/kubernetes/issues/59679

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by bsalamat** [comment]
/hold

Removing the caching could hurt performance. We should add a watch and update the cache when a new default priority class is created.

File: plugin/pkg/admission/priority/admission.go [hunk 1]
```diff
@@ -252,12 +240,5 @@ func (p *PriorityPlugin) getDefaultPriority() (int32, error) {
 	if dpc != nil {
 		priority = dpc.Value
```
Comment (bsalamat) [inline plugin/pkg/admission/priority/admission.go]: here we should just:
```
return dpc.value, nil
```
and at the end of the function
```
return int32(scheduling.DefaultPriorityWhenNoDefaultClassExists), nil
```

**Review by bsalamat — COMMENTED** [review]
/hold cancel

Looks good, just a minor comment.

File: plugin/pkg/admission/priority/admission.go (hunk 1 above)
Comment (wackxu) [reply plugin/pkg/admission/priority/admission.go]: Done

**Comment by bsalamat** [comment]
/assign @liggitt 

Jordan, I am assigning to you since you have the context. Thanks!

**Comment by liggitt** [comment]
removal of the cache looks fine. I'm still unsure why the admission plugin is preventing setting the default on a PriorityClass (it's still possible to get multiple priority classes that both say they are the default, and the behavior of getDefaultPriority is non-deterministic in that case)

**Comment by bsalamat** [comment]
> removal of the cache looks fine. I'm still unsure why the admission plugin is preventing setting the default on a PriorityClass (it's still possible to get multiple priority classes that both say they are the default, and the behavior of getDefaultPriority is non-deterministic in that case)

**Comment by bsalamat** [comment]
> removal of the cache looks fine. I'm still unsure why the admission plugin is preventing setting the default on a PriorityClass (it's still possible to get multiple priority classes that both say they are the default, and the behavior of getDefaultPriority is non-deterministic in that case)

I guess this is not so related to this PR. We can perhaps follow this issue up in a separate issue/PR.
Are you suggesting that this should be done during API validation?

**Comment by liggitt** [comment]
> I guess this is not so related to this PR. We can perhaps follow this issue up in a separate issue/PR.

Yeah, follow up is fine.

> Are you suggesting that this should be done during API validation?

Ensuring a singleton object isn't standard behavior for our API servers today and is racy if done this way.

I'd rather see deterministic behavior when multiple priorityclasses claim to be the default (with multiple defaults, sort by name and take first, etc)

**Comment by bsalamat** [comment]
@liggitt Could you please approve this one if there is no other issue here?

**Comment by bsalamat** [comment]
@deads2k Could you please approve this one as well?

**Comment by bsalamat** [comment]
@deads2k Could you please approve this?

**Pull request merged**